File: Lucy/2025/3-W01/1448_Count_Good_Nodes_in_Binary_Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function goodNodes(root: TreeNode | null): number {
  if (!root) return 0;

  const traverse = (node: TreeNode | null, currentMaxValue: number): void => {
    if (!node) return;

    if (node.val >= currentMaxValue) {
      answer++;
    }

    const newMaxValue = Math.max(currentMaxValue, node.val);

    traverse(node.left, newMaxValue);
    traverse(node.right, newMaxValue);
  };

  let answer = 0;
  traverse(root, root.val);

  return answer;
}

File: Meli/LeetCode 75/1448_Count_Good_Nodes_in_Binary_Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node: TreeNode, max_so_far: int) -> int:
            if not node:
                return 0
            
            is_good = 1 if node.val >= max_so_far else 0
            
            new_max = max(max_so_far, node.val)
            
            return is_good + dfs(node.left, new_max) + dfs(node.right, new_max)
        
        return dfs(root, float('-inf'))